text/0015-Service-Reference-Retrieval.md

```diff
@@ -0,0 +1,233 @@
+        }
+    }
+
+    private:
+    void
+    destroySingleton()
+    {
+        sref.d.Load()->UngetService(b.lock(), true);
+    }
+
+    void
+    destroyPrototype()
+    {
+        auto bundle = b.lock();
+        if (sref)
+        {
+            bool isPrototypeScope
+                = sref.GetProperty(Constants::SERVICE_SCOPE).ToString() == Constants::SCOPE_PROTOTYPE;
+
+            if (isPrototypeScope)
+            {
+                sref.d.Load()->UngetPrototypeService(bundle, interfaceMap);
+            }
+            else
+            {
+                sref.d.Load()->UngetService(bundle, true);
+            }
+        }
+    }
+};
+
+/* @brief Private helper struct used to facilitate the retrieval of a serviceReference from
+    *        a serviceObject.
+    *
+    *        Service consumers can pass a service to the public API ServiceReferenceFromService.
+    *        This method can use the std::get_deleter method to retrieve this object and through
+    *        it the original serviceReference.
+    */
+class CustomServiceDeleter
+{
+    public:
+    CustomServiceDeleter(ServiceHolder<void>* sh) : sHolder(sh) {}
+
+    void
+    operator()(ServiceHolder<void>* sh)
+    {
+        delete sh;
+    }
+
+    [[nodiscard]] ServiceReferenceBase
+    getServiceRef() const
+    {
+        return sHolder->sref;
+    }
+
+    private:
+    ServiceHolder<void> const* const sHolder;
+};
+
+ServiceReferenceU
+ServiceReferenceFromService(std::shared_ptr<void> const& s)
+{
+    auto deleter = std::get_deleter<CustomServiceDeleter>(s);
+    if (!deleter)
+    {
+        throw std::runtime_error("The input is not a CppMicroServices managed ServiceObject");
+    }
+    return deleter->getServiceRef();
+}
+```
+
+The new way that <code>ServiceHolder</code> objects would be constructed can be seen below:
+
+```c++
+// For a Singleton object (from framework/src/Bundle/BundleContext.cpp)
+auto serviceHolder = new ServiceHolder<void>(b, reference, reference.d.Load()->GetService(b.get()), nullptr);
+std::shared_ptr<ServiceHolder<void>> h(serviceHolder, CustomServiceDeleter { serviceHolder });
+return std::shared_ptr<void>(h, h->service.get());
+
+// For a prototype object (from framework/src/service/ServiceObjects.cpp)
+auto sh = new ServiceHolder<void> { bundle_, d->m_reference, nullptr, result };
+std::shared_ptr<ServiceHolder<void>> h(sh, CustomServiceDeleter { sh });
+return InterfaceMapConstPtr(h, h->interfaceMap.get());
+```
+
+## How we teach this
+
+If the proposal is accepted, the CppMicroServices doxygen guide will be modified to reflect the new functionality.
+Most clients will not need this functionality. Configurations can be injected into DS services using ConfigAdmin.
+
+## Drawbacks
+
+- We are using custom deleters in a way that they were not intended to be used.
+- We are writing the code to delete the memory manually rather than allowing <code>std::shared_ptr</code> to do it for us.
```

Comment:
The `shared_ptr` is still managing the memory, and the default deleter would just call `delete sh;` anyway, so I'm not sure this is a real drawback 😃 If we desperately wanted to avoid calling `delete` ourselves, we could do `std::default_delete<ServiceHolder<void>>{}(sh)`, which I'm sure the compiler will optimize to be exactly `delete sh;`.

Reply:
Yes I know that the `std::shared_ptr` is still doing the deletion via the () operator of the deleter. And yes, it would be calling just `delete sh` as we are... However, we are still doing some overriding of the expected deleters of the `std::shared_ptr` which while not unsafe, is just a little strange. 

But yes, I agree this isn't really a drawback. It I guess is moreso just a statement that we aren't actually cusomizing deletion, rather we are using this as a new data storage method. Which I guess falls into the first bullet. So yes, I guess we can delete this line